kimchi/src/folding/mod.rs

```diff
@@ -0,0 +1,223 @@
+use instance_witness::{RelaxableInstance, RelaxablePair};
+use poly_commitment::{commitment::CommitmentCurve, PolyComm, SRS};
+use quadricization::ExtendedWitnessGenerator;
+use std::{fmt::Debug, hash::Hash};
+
+mod error;
+mod expressions;
+mod instance_witness;
+mod quadricization;
+#[cfg(test)]
+mod test;
+
+type Fi<C> = <<C as FoldingConfig>::Curve as AffineCurve>::ScalarField;
+
+pub trait FoldingConfig: Clone + Debug + Eq + Hash + 'static {
+    type Column: FoldingColumnTrait + Debug + Eq + Hash;
+    type Challenge: Clone + Copy + Debug + Eq + Hash;
+    type Curve: CommitmentCurve;
+    type Srs: SRS<Self::Curve>;
+    type Sponge: Sponge<Self::Curve>;
+    type Instance: Instance<Self::Curve>;
+    type Witness: Witness<Self::Curve>;
+    type Structure;
+    type Env: FoldingEnv<
+        <Self::Curve as AffineCurve>::ScalarField,
+        Self::Instance,
+        Self::Witness,
+        Self::Column,
+        Self::Challenge,
+        Structure = Self::Structure,
+    >;
+    fn rows() -> usize;
+}
+
+#[derive(Clone)]
+pub(crate) enum EvalLeaf<'a, F: Field> {
+    Const(F),
+    Col(&'a Vec<F>),
+    Result(Vec<F>),
+}
+
+impl<'a, F: Field> EvalLeaf<'a, F> {
+    fn map<M: Fn(&F) -> F, I: Fn(&mut F)>(self, map: M, in_place: I) -> Self {
+        use EvalLeaf::*;
+        match self {
+            Const(c) => Const(map(&c)),
+            Col(col) => {
+                let res = col.iter().map(map).collect();
+                Result(res)
+            }
+            Result(mut col) => {
+                for cell in col.iter_mut() {
+                    in_place(cell);
+                }
+                Result(col)
+            }
+        }
+    }
+    fn bin_op<M: Fn(&F, &F) -> F, I: Fn(&mut F, &F)>(
+        a: Self,
+        b: Self,
+        map: M,
+        in_place: I,
+    ) -> Self {
+        use EvalLeaf::*;
+        match (a, b) {
+            (Const(a), Const(b)) => Const(map(&a, &b)),
+            (Const(a), Col(b)) | (Col(b), Const(a)) => {
+                let res = b.iter().map(|f| map(f, &a)).collect();
+                Result(res)
+            }
+            (Col(a), Col(b)) => {
+                let res = a.iter().zip(b.iter()).map(|(a, b)| map(a, b)).collect();
+                Result(res)
+            }
+            (Result(mut a), Const(b)) | (Const(b), Result(mut a)) => {
+                for a in a.iter_mut() {
+                    in_place(a, &b);
+                }
+                Result(a)
+            }
+            (Result(mut a), Col(b)) | (Col(b), Result(mut a)) => {
+                for (a, b) in a.iter_mut().zip(b.iter()) {
+                    in_place(a, b);
+                }
+                Result(a)
+            }
+            (Result(mut a), Result(b)) => {
+                for (a, b) in a.iter_mut().zip(b.iter()) {
+                    in_place(a, b);
+                }
+                Result(a)
+            }
+        }
+    }
+    fn unwrap_result(self) -> Vec<F> {
```

Comment:
Since this is the only `unwrap` function, does it make sense to call `unwrap` directly?